ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java

```diff
@@ -493,10 +493,21 @@ public void syncWithSnapshot(long lastSnapshotIndex) {
 
     // Close open log segment if entries are already included in snapshot
     LogSegment openSegment = cache.getOpenSegment();
-    if (openSegment != null && openSegment.getEndIndex() <= lastSnapshotIndex) {
-      fileLogWorker.closeLogSegment(openSegment);
+    long purgeIndex = lastSnapshotIndex;
+    if (openSegment != null && openSegment.hasEntries()) {
+      LOG.debug("Found open segment {}, with end index {}, snapshotIndex {}",
+          openSegment, openSegment.getEndIndex(), lastSnapshotIndex);
+      if (openSegment.getEndIndex() <= lastSnapshotIndex) {
+        fileLogWorker.closeLogSegment(openSegment);
+        cache.rollOpenSegment(true);
+        if (openSegment.getEndIndex() == lastSnapshotIndex) {
+          // Since purgeImpl does not delete any segment which has overlap,
+          // passing in snapshotIndex + 1.
+          purgeIndex = lastSnapshotIndex + 1;
+        }
```

Comment:
Can we add this fix in LogSegmentList#purge function? If the endIndex of last segment is same as purge index, we can remove it.

Reply:
Done.